eclair-core/src/main/scala/fr/acinq/eclair/payment/PaymentRequest.scala

```diff
@@ -377,41 +390,58 @@ object PaymentRequest {
 
     val dataLengthCodec: Codec[Long] = uint(10).xmap(_ * 5, s => (s / 5 + (if (s % 5 == 0) 0 else 1)).toInt)
 
-    def dataCodec[A](valueCodec: Codec[A]): Codec[A] = paddedVarAlignedBits(dataLengthCodec, valueCodec, multipleForPadding = 5)
+    def dataCodec[A <: TaggedField](valueCodec: Codec[A]): Codec[A] = paddedVarAlignedBits(dataLengthCodec, valueCodec, multipleForPadding = 5)
+
+    def dataCodecLengthDiscriminated[A <: TaggedField](length: Long, valueCodec: Codec[A]): Codec[A] = paddedVarAlignedBits(provide(length), valueCodec, multipleForPadding = 5)
+
+    val paymentHashCodec: Codec[PaymentHashTag] = discriminatorWithDefault(
+      discriminated.by(dataLengthCodec)
+        .typecase(260, dataCodecLengthDiscriminated(260, bytes32.as[PaymentHash])),
+      bits.as[InvalidPaymentHash].upcast[PaymentHashTag])
+
+    val paymentSecretCodec: Codec[PaymentSecretTag] = discriminatorWithDefault(
+      discriminated.by(dataLengthCodec)
+        .typecase(260, dataCodecLengthDiscriminated(260, bytes32.as[PaymentSecret])),
+      bits.as[InvalidPaymentSecret].upcast[PaymentSecretTag])
+
+    val descriptionHashCodec: Codec[DescriptionHashTag] = discriminatorWithDefault(
+      discriminated.by(dataLengthCodec)
+        .typecase(260, dataCodecLengthDiscriminated(260, bytes32.as[DescriptionHash])),
+      bits.as[InvalidDescriptionHash].upcast[DescriptionHashTag])
 
```

Comment:
A comment that explains where `260` comes from (maybe a quote from BOLT11) may help here